jl4-core/src/L4/Lexer.hs

```diff
@@ -257,44 +259,73 @@ directives =
   , (TCheckDirective,      "CHECK")
   ]
 
-integerLiteral :: Lexer (Text, Int)
+integerLiteral :: Lexer (Text, Integer)
 integerLiteral =
       (\ x (xs, i) -> (x <> xs, negate i)) <$> string "-" <*> decimal
   <|> decimal
 
-decimal :: Lexer (Text, Int)
+decimal :: Lexer (Text, Integer)
 decimal = decimal_ <?> "integer"
 
-decimal_ :: Lexer (Text, Int)
+decimal_ :: Lexer (Text, Integer)
 decimal_ = (\ s -> (s, mkNum s)) <$> takeWhile1P (Just "digit") isDigit
   where
-    mkNum :: Text -> Int
+    mkNum :: Text -> Integer
     mkNum = foldl' step 0 . chunkToTokens (Proxy :: Proxy Text)
     step a c = a * 10 + fromIntegral (digitToInt c)
 
+rationalLiteral :: Lexer (Text, Rational)
+rationalLiteral =
+      (\ x (xs, i) -> (x <> xs, negate i)) <$> string "-" <*> floatP
+  <|> floatP
+{-# INLINE rationalLiteral #-}
+
+floatP :: Lexer (Text, Rational)
+floatP = float_ <?> "float"
+{-# INLINE floatP #-}
+
+float_ :: Lexer (Text, Rational)
+float_ = do
+  (t, c') <- decimal_
+  (c, e, t2) <- dotDecimal_ c'
+  pure (t <> "." <> t2, realToFrac $ Sci.scientific c e)
+{-# INLINE float_ #-}
+
+dotDecimal_ :: Integer -> Lexer (Integer, Int, Text)
+dotDecimal_ c' = do
+  void (satisfy (== '.'))
+  let mkNum = foldl' step (c', 0) . chunkToTokens (Proxy :: Proxy Text)
+      step (a, e') c =
+          ( (a * 10 + fromIntegral (digitToInt c))
+          , (e' - 1)
+          )
+  (\ s -> let (c, e) = mkNum s in (c, e, s)) <$> takeWhile1P (Just "digit") isDigit
+{-# INLINE dotDecimal_ #-}
+
```

Comment:
is there a reason why we cannot use [one of the combinators provided by megaparsec](https://hackage.haskell.org/package/megaparsec-9.7.0/docs/Text-Megaparsec-Char-Lexer.html#v:scientific)?

Reply:
Yes, we lose the original characters